aptos-move/framework/aptos-framework/sources/code.move

```diff
@@ -0,0 +1,183 @@
+/// This module supports functionality related to code management.
+module aptos_framework::code {
+    use std::string::String;
+    use std::error;
+    use std::signer;
+    use std::vector;
+
+    // ----------------------------------------------------------------------
+    // Code Publishing
+
+    /// The package registry at the given address.
+    struct PackageRegistry has key {
+        /// Packages installed at this address.
+        packages: vector<PackageMetadata>,
+    }
+
+    /// Metadata for a package.
+    struct PackageMetadata has store, copy, drop {
+        /// Name of this package.
+        name: String,
+        /// The upgrade policy of this package.
+        upgrade_policy: UpgradePolicy,
+        /// The list of modules installed by this package.
+        modules: vector<ModuleMetadata>,
+        /// Address aliases which where used when the modules above were compiled.
+        address_aliases: vector<AddressAlias>,
+        /// Dependencies which were used when this package was compiled.
+        deps: vector<PackageDep>,
+    }
+
+    /// Metadata about a module in a package.
+    struct ModuleMetadata has store, copy, drop {
+        /// Name of the module.
+        name: String,
+        /// Source map, in internal encoding
+        source_map: vector<u8>,
+        /// Source text.
+        source: String,
+    }
+
+    /// A package dependency. This consists of an address and the name of a package
+    /// at this address.
+    struct PackageDep has store, copy, drop {
+        addr: address,
+        name: String,
+    }
+
+    /// An address alias.
+    struct AddressAlias has store, copy, drop {
+        alias: String,
+        addr: address
+    }
+
+    /// Describes an upgrade policy
+    struct UpgradePolicy has store, copy, drop {
+        policy: u8
+    }
+
+    /// A package is attempted to publish with module names clashing with modules published by other packages on this
+    /// address.
+    const EMODULE_NAME_CLASH: u64 = 0x1;
+
+    /// A package is attempted to upgrade which is marked as immutable.
+    const EUPGRADE_IMMUTABLE: u64 = 0x2;
+
+    /// A package is attempted to upgrade with a weaker policy than previously.
+    const EUPGRADE_WEAKER_POLICY: u64 = 0x3;
+
+    /// Whether unconditional code upgrade with no compatibility check is allowed. This
+    /// publication mode should only be used for modules which aren't shared with user others.
+    /// The developer is responsible for not breaking memory layout of any resources he already
+    /// stored on chain.
+    public fun upgrade_policy_no_compat(): UpgradePolicy {
+        UpgradePolicy{policy: 0}
```

Comment:
nit: Can we define these 0/1 numbers as constants?

Reply:
Not sure what this buys us. Constants are only local in Move, unfortunately, and so we will always use those functions to denote policies.

Reply:
Yeah enum support will be amazing! But I guess we might not have time to do that before Mainnet.